corehq/ex-submodules/casexml/apps/case/tests/test_extract_caseblocks.py

```diff
@@ -90,6 +92,37 @@ def test_blocks_in_repeat(self):
             self.assertEqual(['data', 'parent', 'repeats', 'group'], blocks_back[i].path)
 
 
+class TempCaseBlockCacheTests(TestCase):
+    def setUp(self):
+        self.domain = 'test_case_block_cache'
+        self.form_id = '1234'
+
+    def test_no_db_hit_if_cached(self):
+        cache_obj = TempCaseBlockCache()
+        blocks = extract_case_blocks({
+            'data': {
+                'some': 'stuff'
+            },
+            'case': {CASE_ATTR_ID: uuid.uuid4().hex}
+        })
+        cache_obj.cache[self.form_id] = blocks
+        block_form = create_form_for_test(self.domain, form_id=self.form_id)
```

Comment:
SQL database interactions in tests are normally wrapped a transaction, which is then rolled back at the end of each test to reset the database to a clean state. However, forms and cases and a few other models are stored in "shard" databases, which do not use transactions. Therefore it is necessary to do explicit test cleanup using something like `self.addCleanup(block_form.delete)`.